lib/restrictor_mma.in.cu

```diff
@@ -0,0 +1,279 @@
+        tp.set_max_shared_bytes = true;
+        launch_cuda<RestrictorMma>(tp, stream, arg);
+      } else {
+        errorQuda("Using too many shared memory bytes per block: %d", shared_bytes);
+      }
+    }
+
+    void launch_mma(TuneParam &tp, const qudaStream_t &stream) { expand.expand(tp, stream); }
+
+    void apply(const qudaStream_t &stream)
+    {
+      TuneParam tp = tuneLaunch(*this, getTuning(), getVerbosity());
+      launch_mma(tp, stream);
+    }
+  };
+
+  template <typename store_t, typename in_t, int fineSpin, int fineColor, int coarseColor, int nVec>
+  void RestrictMma(ColorSpinorField &out, const ColorSpinorField &in, const ColorSpinorField &v,
+                   const int *fine_to_coarse, const int *coarse_to_fine, const int *const *spin_map, int parity)
+  {
+    if (out.Nspin() != 2) errorQuda("Unsupported nSpin %d", out.Nspin());
+    constexpr int coarseSpin = 2;
+
+    // first check that the spin_map matches the spin_mapper
+    spin_mapper<fineSpin, coarseSpin> mapper;
+    for (int s = 0; s < fineSpin; s++)
+      for (int p = 0; p < 2; p++)
+        if (mapper(s, p) != spin_map[s][p]) errorQuda("Spin map does not match spin_mapper");
+
+    if (v.Precision() == QUDA_HALF_PRECISION) {
+      if constexpr (is_enabled(QUDA_HALF_PRECISION)) {
+        RestrictMmaLaunch<store_t, in_t, short, fineSpin, fineColor, coarseSpin, coarseColor, nVec> restrictor(
+          out, in, v, fine_to_coarse, coarse_to_fine, parity);
+      } else {
+        errorQuda("QUDA_PRECISION=%d does not enable half precision", QUDA_PRECISION);
+      }
+    } else if (v.Precision() == in.Precision()) {
+      RestrictMmaLaunch<store_t, in_t, store_t, fineSpin, fineColor, coarseSpin, coarseColor, nVec> restrictor(
+        out, in, v, fine_to_coarse, coarse_to_fine, parity);
+    } else {
+      errorQuda("Unsupported V precision %d", v.Precision());
+    }
+  }
+
+  template <typename store_t, int fineColor, int coarseColor, int nVec>
+  void RestrictMma(ColorSpinorField &out, const ColorSpinorField &in, const ColorSpinorField &v,
+                   const int *fine_to_coarse, const int *coarse_to_fine, const int *const *spin_map, int parity)
+  {
+    if (!is_enabled_spin(in.Nspin())) errorQuda("nSpin %d has not been built", in.Nspin());
+
+    if (in.Nspin() == 2) {
+      RestrictMma<store_t, store_t, 2, fineColor, coarseColor, nVec>(out, in, v, fine_to_coarse, coarse_to_fine,
+                                                                     spin_map, parity);
+    } else if constexpr (fineColor == 3) {
+      if (in.Nspin() == 4) {
+        if constexpr (is_enabled_spin(4)) {
+          if (in.Precision() == out.Precision()) {
+            RestrictMma<store_t, store_t, 4, fineColor, coarseColor, nVec>(out, in, v, fine_to_coarse, coarse_to_fine,
+                                                                           spin_map, parity);
+          } else if (in.Precision() == QUDA_HALF_PRECISION) {
+#if 0
```

Comment:
Can you add some documentation here, describing why this is disabled at the moment?

Reply:
https://github.com/lattice/quda/pull/1497/commits/87d332f5178d97dc15e642d362f1e5d4e1f2bfd8